Shared/ModuleHandle.ps1

```diff
@@ -0,0 +1,85 @@
+﻿# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+
+<#
+.SYNOPSIS
+This script defines a function `Request-Module` that checks for the presence of specified PowerShell modules.
+If a module is not found, it attempts to install it on current user scope.
+The function accepts a list of module names and an optional minimum version for the modules.
+It returns a boolean indicating whether all specified modules are installed successfully.
+
+.PARAMETER Modules
+ Mandatory array of strings specifying the names of the modules to check and install if necessary.
+.PARAMETER MinModuleVersion
+ Optional parameter to specify the minimum version of the modules (default is null).
+
+.OUTPUTS
+bool. A boolean indicating whether all specified modules are installed successfully.
+
+.EXAMPLE
+$requestModule = Request-Module -Modules "ExchangeOnlineManagement"
+This example checks if the "ExchangeOnlineManagement" module is installed. If it is not found, the script attempts to install it.
+
+.EXAMPLE
+$requestModule = Request-Module -Modules "ExchangeOnlineManagement" -MinModuleVersion $MinModuleVersion
+This example checks if the "ExchangeOnlineManagement" module with a specified minimum version is installed. If it is not found, the script attempts to install it.
+#>
+
+function Request-Module {
+    [OutputType([bool])]
+    param (
+        [Parameter(Mandatory = $true)]
+        [string[]]$Module,
+        [Parameter(Mandatory = $false)]
+        [System.Version]$MinModuleVersion = $null
+    )
+
+    $noFoundError = $true
+    foreach ($m in $Module) {
+        Write-Verbose "Checking $m PowerShell Module"
+        $getParams = @{
+            Name        = $m
+            ErrorAction = 'SilentlyContinue'
+        }
+        if ($MinModuleVersion) {
+            $getParams["MinimumVersion"] = $MinModuleVersion
+            Write-Verbose "with minimum version $minModuleVersion"
+        } else {
+            Write-Verbose "without minimum version"
+        }
+        $installed = Get-InstalledModule @getParams
```

Comment:
We run the `Get-InstalledModule` multiple times. Can we run it just one time to query information about all available modules before we enter the `foreach`, cache the information and then continue processing based on the cached information within the `foreach`?

Reply:
It should work, I will test it. Whit the current script we will check every time only will be better if we request the same module or in similar situation. Looks better your suggestion.